Report all changes in calls to rebuild and update

airhorns opened the pull request:
Before this change, callers of the binary client know how many things changed, but not which things changed. There's a few bits of business logic gadget side that would be a lot easier if they knew exactly what changed, namely looking for changes to package.json during operation-runner commands, or looking for changes only to client-side files so we don't need to restart the server. This hands that information back to the client so they can work with it. We were already generating it so no huge performance hit there, but it will potentially be big when we send it back to the client so I think we may need to use this carefully.

scott-rc commented:
This is probably going to have merge conflicts with #34 but I'll fix them once it's merged.

airhorns committed:
Make the clients return a list of the files that changed after an update

Before this change, callers of the binary client know how many things changed, but not which things changed. There's a few bits of business logic gadget side that would be a lot easier if they knew exactly what changed, namely looking for changes to package.json during operation-runner commands, or looking for changes only to client-side files so we don't need to restart the server. This hands that information back to the client so they can work with it. We were already generating it so no huge performance hit there, but it will potentially be big when we send it back to the client so I think we may need to use this carefully.

angelini commented on pkg/cli/rebuild.go:
```diff
@@ -51,10 +52,14 @@ func NewCmdRebuild() *cobra.Command {
 					key.Directory.Field(dir),
 					key.Version.Field(version),
-					key.DiffCount.Field(count),
+					key.DiffCount.Field(uint32(len(diff.Updates))),
```
The segfault you are seeing in the JS tests is from this line.

This won't segfault if `diff.Updates` is `nil` but it will segfault if `diff` (which is a pointer) is `nil`.

angelini commented:
There's also a tool in `fsdiff` for writing diff files to disk if that would be better than writing them to stdout:

https://github.com/gadget-inc/fsdiff/blob/main/pkg/diff/writer.go#L34-L40

angelini commented on test/client_combined_test.go:
```diff
@@ -194,7 +194,7 @@ func TestCombinedWithChangingObjectTypes(t *testing.T) {
 	rebuild(tc, c, 1, i(2), tmpDir, nil, expectedResponse{
 		version: 2,
-		count:   3,
+		count:   5,
```
Do you know why all of the file counts changed?

airhorns replied:
I think but I am not sure that it's the full summary emitting more removes than were detected before. When a file becomes a folder for example, we get a `{path: "a", op: "REMOVE"}` in addition to a `{path: "a/", op: "ADD"}` that I don't think we got before without doing the full diff. I didn't log them out though, I can prove that's the case.

airhorns committed:
Add an option to the binary client to report if any files matching a glob have changed

Instead of listing out all the files and checking globs in JS land, we may as well pass the globs down to DL where it can do it faster and potentially eventually more efficiently.

airhorns commented:
Updated this to also accept some globs as an option, and to check those globs against all the files that changed in memory, instead of logging them all out just to do the same thing in node land.

airhorns commented:
ping @scott-rc and @angelini , this is blocking for frontend reloading being good so would love to get it out early next week, which means into gadget sooner than that 💖

scott-rc commented on pkg/cli/rebuild.go:
```diff
@@ -64,10 +78,54 @@ func NewCmdRebuild() *cobra.Command {
+	return false, nil
+}
+
+func LogIfGlobsMatched(globs []string, diff *fsdiff_pb.Diff) error {
+	if len(globs) > 0 {
+		matched, err := anyMatchedGlobs(globs, diff)
+		if err != nil {
+			return err
+		}
+		if matched {
+			fmt.Printf("GLOB-MATCH: files matching globs were matched during rebuild\n")
+		}
+	}
+	return nil
+}
```
Aren't we trying to figure out if **only** specific files changed? This only tells us if specific files changed, not if they only changed. Unless I'm thinking about this wrong?

scott-rc commented on js/src/binary-client.ts:
```diff
@@ -249,4 +293,40 @@ export class DateiLagerBinaryClient {
     return subprocess;
   }
+
+  /**
+   *
+   * @param stream
+   * @param options
+   * @param options.listUpdated
+   * @param options.checkGlobs
+   */
```
You can use this doc comment so that ESLint doesn't complain about missing docs and TypeScript doesn't emit types for it.

```suggestion
  /** @internal */
```

angelini commented on js/src/binary-client.ts:
```diff
@@ -136,35 +177,40 @@ export class DateiLagerBinaryClient {
       },
       async () => {
-        const result = await this._call("update", project, directory, ["--dir", directory], options);
+        const args = ["--dir", directory, `--log-updates=${String(!!options?.listUpdated)}`, String(!!options?.listUpdated)];
```
Why is the `options.listUpdated` arg included twice?

angelini commented:
This change unfortunately removes an existing optimization: `summary: false`.

Since the sandboxes are read only, they receive updates from DL but never need to push them back upstream. Because of this the sandbox could avoid summarizing the data on disk as it would never need to generate a diff file.

I'm not convinced that building a diff and summary on disk - our current bottleneck - is faster than just asking DL if frontend files have changed between two versions.

This seems like it might be a performance regression and it complicates the CLI args in order to avoid an extra call to DL. I don't have the performance numbers in front of me but disk bound operations on the sandbox have always been slower than PG operations in DL.
